File: 0213-house-robber-ii/0213-house-robber-ii.java
class Solution {
    public int rob(int[] nums) {
        int arr[]=new int [nums.length];
        if(nums.length==1) return nums[0];
        if (nums.length == 2) return Math.max(nums[0], nums[1]);
        arr[0]=nums[0];
        arr[1]=Math.max(arr[0],nums[1]);
        for(int i=2;i<nums.length-1;i++)
            arr[i]=Math.max(arr[i-1],nums[i]+arr[i-2]);
        int dp[]=new int[nums.length];
        dp[1]=nums[1];
        dp[2]=Math.max(dp[1],nums[2]);
        for(int i=3;i<nums.length;i++)
            dp[i]=Math.max(dp[i-1],nums[i]+dp[i-2]);
        return Math.max(arr[nums.length-2],dp[nums.length-1]);
    }
}

File: 0213-house-robber-ii/0213-house-robber-ii.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
      if (nums.size() == 1) return nums[0];
        int n = nums.size();
        vector<int> DP (n, 0);
        vector<int> new_DP (n,0);
        DP[0] = 0; // dont rob 1st house.
        DP[1] = nums[1];
        new_DP[0] = nums[0]; // dont rob last house.
        new_DP[1] = max(nums[0] , nums[1]);
        for (int i=2; i<nums.size(); i++)
            DP[i] = max(DP[i-1], nums[i]+ DP[i-2]);
        for (int i=2; i<nums.size(); i++)
        {
            if (i== n-1) new_DP[i] = new_DP[i-1]; // dont rob last house.
            else new_DP[i] = max(new_DP[i-1], nums[i]+ new_DP[i-2]);
        }
        return max(DP.back(), new_DP.back());  
    }
};